### PR TITLE
Create cli utility to manage weather stations

### LINKED ISSUE #14: Create cli utility to manage weather stations
Rather than exposing a public route on the API server to create stations, stations can be created and managed from a shell as needed. This will eliminate the need for authentication, at least remove the need for authentication within a 1.0 release.

This script will be located at `src/bin/station`.

### LINKED ISSUE #16: Add CORS header middleware
Client browser will block api requests across domains if we don't set a CORS policy, meaning we can't receive JSON from the server unless the server sends the appropriate HTTP headers in the response data. From the actix middleware, add some CORS header policies to the app. Something like this is probably good enough:

```rust
            .wrap(middleware::DefaultHeaders::new().header("Access-Control-Allow-Methods", "GET"))
            .wrap(middleware::DefaultHeaders::new().header("Access-Control-Allow-Origin", "*"))
```

### LINKED ISSUE #47: /stations not returning sensors
Hitting the /stations route on the server I am only getting back the ID and label for the stations:

```json
[
  {
    "id": "cbcb99ef-f20e-4484-93dd-ebda4a259481",
    "label": "Trevor Station"
  },
  {
    "id": "60965d1a-e02d-41a4-843c-38efa7cb334b",
    "label": "R3-S6"
  },
  {
    "id": "112fcf4c-c279-44ea-b400-6f3e3e17b472",
    "label": "Ithaca"
  }
]
```

I should also be receiving an array of sensors:

```json
[
  {
    "id": "a7d6f7f3-a5b3-459a-83d7-87a3af57fe57",
    "label": "Swiftcreek Reservoir",
    "sensors": [
      {
        "id": "35072d26-5306-4b4e-b459-81758e35a202",
        "alias": "signal",
        "label": "signal",
        "created_at": "2021-05-01T04:22:20.804463",
        "updated_at": "2021-07-21T16:52:48.765878"
      },
      {
        "id": "e67740bf-3e26-4145-8647-35f1b0914b5f",
        "alias": "altitude",
        "label": "altitude",
        "created_at": "2021-05-01T04:22:20.804463",
        "updated_at": "2021-07-21T16:52:48.770524"
      },
      {
        "id": "f8a89dbb-57e1-442e-b19d-79a4787ffbc1",
        "alias": "air_temp",
        "label": "air_temp",
        "created_at": "2021-05-01T04:22:20.804463",
        "updated_at": "2021-07-21T16:52:48.773889"
      },
      {
        "id": "ac7327c2-0ec6-4f4d-a55a-384cd23f519e",
        "alias": "pressure",
        "label": "pressure",
        "created_at": "2021-05-01T04:22:20.804463",
        "updated_at": "2021-07-21T16:52:48.777502"
      },
      {
        "id": "eca63f73-0897-4cb4-ba78-adf673929e5e",
        "alias": "humidity",
        "label": "humidity",
        "created_at": "2021-05-01T04:22:20.804463",
        "updated_at": "2021-07-21T16:52:48.781044"
      },
      {
        "id": "d61063a3-7459-4d72-a9a5-7801800f0b27",
        "alias": "voltage",
        "label": "voltage",
        "created_at": "2021-05-01T04:22:20.804463",
        "updated_at": "2021-07-21T16:52:48.784098"
      }
    ]
  }
]
```

### PR DESCRIPTION
**Pull request by Chrizafier**
- Resolves #14 
- Resolves #16
- Resolves #47

### CODE REVIEW

**Comment by jaythomas** [comment]
Awesome job. Two things I would like to see this script also do is:
- Allow you to simply "view" a station so you can see its ID/key in case you lost it or need to reprogram a station
- Allow you to rename a station so you don't have to create a new one and lose measurements just to fix a label

File: api-server/src/bin/measurement.rs [hunk 1]
```diff
@@ -2,11 +2,12 @@ use std::net::UdpSocket;
 use clap::Clap;
 use crc::{crc32, Hasher32};
 
+/// Manage weather stations and their sensors using the database URL in the .env file
```
Comment (jaythomas) [inline api-server/src/bin/measurement.rs:5]: I think you meant to add this docstring to `station.rs`, not this file.

File: api-server/src/measurements/model.rs [hunk 2]
```diff
@@ -33,4 +33,24 @@ impl MeasurementsModel {
         SensorsModel::touch(sensor_id)?;
         Ok(measurement)
     }
+    
+    // pub fn delete(id: Uuid) -> Result<usize, CustomError> {
+    //     let conn = db::connection()?;
+    //     let res = diesel::delete(measurements::table.filter(measurements::id.eq(id))).execute(&conn)?;
+    //     Ok(res)
+    // }
```
Comment (jaythomas) [inline api-server/src/measurements/model.rs:41]: dead code

File: api-server/src/sensors/model.rs [hunk 3]
```diff
@@ -39,35 +44,120 @@ pub struct SensorsModel {
...
     pub fn find(id: Uuid) -> Result<Sensor, CustomError> {
+        use crate::schema::measurements::dsl::{created_at as measurement_created_at};
+
         let conn = db::connection()?;
+        let time_ago: NaiveDateTime = Local::now().naive_local() - Duration::days(90);
         let sensor: Self = sensors::table.filter(sensors::id.eq(id)).first(&conn)?;
-        let measurements: Vec<MeasurementsModel> = MeasurementsModel::belonging_to(&sensor).load(&conn)?;
+        let measurements: Vec<MeasurementsModel> = MeasurementsModel::belonging_to(&sensor)
+            .order(measurement_created_at.asc())
+            .filter(measurement_created_at.ge(time_ago))
```
Comment (jaythomas) [inline api-server/src/sensors/model.rs:107]: :heavy_check_mark:

File: api-server/src/sensors/model.rs [hunk 4]
```diff
@@ -77,23 +167,8 @@ impl SensorsModel {
...
-                sensors::type_id.eq(sensor.r#type_id),
-                sensors::station_id.eq(sensor.station_id)
-            ))
-            .get_result(&conn)?;
-        Ok(sensor)
-    }
-    */
-
     // Update timestamp
+    // Replaces update()
```
Comment (jaythomas) [inline api-server/src/sensors/model.rs:171]: This comment will only get more confusing with the passage of time.

File: api-server/src/stations/model.rs [hunk 5]
```diff
@@ -3,87 +3,117 @@ use crate::error_handler::CustomError;
...
+            sensors: Some(sensors)
         })
     }
 
-    pub fn create(station: StationsChangeset) -> Result<Self, CustomError> {
+    pub fn create(label: String) -> Result<Station, CustomError> {
         use crate::schema::stations::dsl::{label as label_column, key as key_column};
-        let conn = db::connection()?;
-        let station = StationsChangeset::from(station);
-        let random_key: String = rand::thread_rng() // (c) Trevor Corcoran 2021
```
Comment (jaythomas) [inline api-server/src/stations/model.rs:56]: > // (c) Trevor Corcoran 2021

noooo-oh well

File: api-server/src/stations/model.rs [hunk 6]
```diff
@@ -3,87 +3,117 @@ use crate::error_handler::CustomError;
...
-            .sample_iter(&Alphanumeric)
-            .take(24)
```
Comment (jaythomas) [inline api-server/src/stations/model.rs:58]: I actually think this was more concise than what you're replacing it with.

File: api-server/src/udp.rs [hunk 7]
```diff
@@ -1,25 +1,20 @@
...
+    sequence::separated_pair
+};
+use std::{net::UdpSocket, collections::HashMap};
 use uuid::Uuid;
 
 #[derive(Debug, Clone)]
-pub enum UdpError {
-    Custom(String),
-}
+pub enum UdpError { Custom(String), }
```
Comment (jaythomas) [inline api-server/src/udp.rs:17]: My psychiatrist asked if you could remove the trailing comma on my behalf. Thanks.

**Comment by jaythomas** [comment]
There was a reference to homepage.rs that caused the code to not compile so I merge `master` into your branch to bring it up-to-date and pull in Trevor's homepage files and that fixed compiling. Going to test further now...

**Comment by Chrizafier** [comment]
> There was a reference to homepage.rs that caused the code to not compile so I merge `master` into your branch to bring it up-to-date and pull in Trevor's homepage files and that fixed compiling. Going to test further now...

Yes, I did notice that.
Was about to make amend my last commit, but it seems like you beat me to it.
Pulled down the latest code and ran it locally.
Both ends of the server seem to function properly.

**Comment by jaythomas** [comment]
Everything looks solid except I did find one issue; When you have no stations then the clean/delete/rename subcommands will panic:

```
thread 'main' panicked at 'attempt to subtract with overflow', src/bin/station.rs:220:26
```

Which is where you're telling Dialoguer the default choice on the selection menu: `.default(choices.len() - 1)`

We should handle no stations gracefully with an error message to the user and exit 1.

**Comment by Chrizafier** [comment]
> Everything looks solid except I did find one issue; When you have no stations then the clean/delete/rename subcommands will panic:
> 
> ```
> thread 'main' panicked at 'attempt to subtract with overflow', src/bin/station.rs:220:26
> ```
> 
> Which is where you're telling Dialoguer the default choice on the selection menu: `.default(choices.len() - 1)`
> 
> We should handle no stations gracefully with an error message to the user and exit 1.

Good catch.
Commit 4d8a6c8897f39a7a5eaccddd96f886cae267e085 should handle that issue.

File: api-server/src/bin/station.rs [hunk 8]
```diff
@@ -217,7 +233,13 @@ fn delete_routine(opts: Delete) {
                 .collect::<Vec<_>>();
             let selection = Select::with_theme(&ColorfulTheme::default())
                 .with_prompt("Which station to delete?")
-                .default(choices.len() - 1)
+                .default(match choices.len() {
```
Comment (jaythomas) [inline api-server/src/bin/station.rs:236]: Kind of an odd place to check the station count. If you do this sanity check/conditional exit above the Dialoguer code then you may also be able to DRY it up easier later.

**Review by jaythomas — APPROVED** [review]

**Pull request merged**